File: 0452-minimum-number-of-arrows-to-burst-balloons/0452-minimum-number-of-arrows-to-burst-balloons.java
class Solution {
    public int findMinArrowShots(int[][] points) {
        //common template for overlapping ends
        int numberOfOverlappingEnds = 0;
        int minimumEnd = Integer.MAX_VALUE;
        
        Arrays.sort(points, Comparator.comparingInt(o -> o[1]));
        
        for (int[] p : points) {
            
            if (p[0] > minimumEnd) {
                numberOfOverlappingEnds++;
                minimumEnd = p[1];
            } else {
                minimumEnd = Math.min(minimumEnd, p[1]);
            }
        }
        return numberOfOverlappingEnds + (points.length > 0 ? 1 : 0);

    }
}

File: 0452-minimum-number-of-arrows-to-burst-balloons/0452-minimum-number-of-arrows-to-burst-balloons.py
class Solution:
    def findMinArrowShots(self, P: List[List[int]]) -> int:
        an=1
        P=sorted(P,key=lambda x: x[0])
        start,end=P[0][0],P[0][1]
 
        for i in range(1,len(P)):
            if P[i][0]>end:
                an+=1
                end=P[i][1]
            else:
                end = min(end, P[i][1])
        return an
    